x/stakeibc/keeper/msg_server_redeem_stake.go

```diff
@@ -109,7 +121,10 @@ func (k msgServer) RedeemStake(goCtx context.Context, msg *types.MsgRedeemStake)
 		return nil, errorsmod.Wrapf(types.ErrInvalidHostZone, "host zone not found in unbondings: %s", hostZone.ChainId)
 	}
 	hostZoneUnbonding.NativeTokenAmount = hostZoneUnbonding.NativeTokenAmount.Add(nativeAmount)
-	hostZoneUnbonding.UserRedemptionRecords = append(hostZoneUnbonding.UserRedemptionRecords, userRedemptionRecord.Id)
+	if !userHasRedeemedThisEpoch {
```

Comment:
Can you explain how `userHasRedeemedThisEpoch` tells us whether the URR was already appended to the HZU? My understanding is that it tells us whether the URR exists (as it is the _found_ return value from `GetUserRedemptionRecord()`.

Reply:
I think this is using an assumption that when the URR is created, it's always appended to the HZU. Might be good to double check that assumption is valid, but I think it's valid!

Reply:
```
userRedemptionRecord, userHasRedeemedThisEpoch := k.RecordsKeeper.GetUserRedemptionRecord(ctx, redemptionId)
```

`userHasRedeemedThisEpoch` is true if a URR is found for the `redemptionId`, which is a key composed of chain id, epoch number (day), sender addr.

Before a user has redeemed, no record will be found. After a user has redeemed in an epoch (as defined by the key), the URR will be stored and found.

Reply:
I don't think it has anything to do with the HZU - although @shellvish I might be missing something, can you double check this?

Reply:
I believe Riley is referring to [these lines](https://github.com/Stride-Labs/stride/blob/30ee082fc59c420367fe922bfced8c46fdf8aa2a/x/stakeibc/keeper/msg_server_redeem_stake.go#L124-L127C3) where we append the redemption record _id_ onto the HZU record.

And I agree with Vishal that it's just based on the assumption that we always append the ID to the HZU during the first redemption! I've confirmed this as well - in fact, as far as I can tell, the place this list is actually used is in the HZU query.